applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp

```diff
@@ -583,6 +587,71 @@ class EigensolverStrategy
+     *  Beware that this implementation is only valid for Block B&S, since the master-slave constraints
+     *  don't work with Elimination B&S yet.
+     */
+    void ReconstructSolution(
+        DenseMatrixType& rEigenvectors
+    )
+    {
+        KRATOS_TRY
+
+        auto& r_model_part = BaseType::GetModelPart();
+        const std::size_t number_of_eigenvalues = rEigenvectors.size1();
+
+        struct TLS{
+            Matrix relation_matrix;
+            Vector constant_vector;
+            Vector master_dofs_values;
+        };
+
+        for (std::size_t i_eigenvalue = 0; i_eigenvalue < number_of_eigenvalues; ++i_eigenvalue){
+
+            // Reset slave dofs
+            block_for_each(r_model_part.MasterSlaveConstraints(), [&i_eigenvalue, &rEigenvectors](const MasterSlaveConstraint& r_master_slave_constraint){
+                const auto& r_slave_dofs_vector = r_master_slave_constraint.GetSlaveDofsVector();
+                for (const auto& r_slave_dof: r_slave_dofs_vector){
+                    rEigenvectors(i_eigenvalue, r_slave_dof->EquationId()) = 0.0;
+                }
+            });
+
+            // Apply constraints
+            block_for_each(r_model_part.MasterSlaveConstraints(), TLS(), [&i_eigenvalue, &rEigenvectors, &r_model_part](const MasterSlaveConstraint& r_master_slave_constraint, TLS& rTLS){
+                // Detect if the constraint is active or not. If the user did not make any choice the constraint
+                // It is active by default
+                bool constraint_is_active = true;
+                if (r_master_slave_constraint.IsDefined(ACTIVE))
+                    constraint_is_active = r_master_slave_constraint.Is(ACTIVE);
+                if (constraint_is_active) {
+                    // Saving the master dofs values
+                    const auto& r_master_dofs_vector = r_master_slave_constraint.GetMasterDofsVector();
+                    const auto& r_slave_dofs_vector = r_master_slave_constraint.GetSlaveDofsVector();
+                    rTLS.master_dofs_values.resize(r_master_dofs_vector.size());
+                    for (IndexType i = 0; i < r_master_dofs_vector.size(); ++i) {
+                        rTLS.master_dofs_values[i] = rEigenvectors(i_eigenvalue, r_master_dofs_vector[i]->EquationId());
+                    }
+                    // Apply the constraint to the slave dofs
+                    r_master_slave_constraint.GetLocalSystem(rTLS.relation_matrix, rTLS.constant_vector, r_model_part.GetProcessInfo());
+                    double aux;
+                    for (IndexType i = 0; i < rTLS.relation_matrix.size1(); ++i) {
+                        aux = rTLS.constant_vector[i];
+                        for(IndexType j = 0; j < rTLS.relation_matrix.size2(); ++j) {
+                            aux += rTLS.relation_matrix(i,j) * rTLS.master_dofs_values[j];
+                        }
+                        #pragma omp atomic
```

Comment:
added atomic update here too

Reply:
please use [this](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/atomic_utilities.h#L39) instead

Reply:
done